### ISSUE TITLE
Bonding wired and wireless interfaces, connecting only to a WiFi network resolves in no internet connection at all

### ISSUE THREAD

**Issue by fmore-xyz** [systemd/systemd#22830]
This is a clone of https://github.com/systemd/systemd/issues/9011 where an eth and a wifi interfaces are bonded together and when only the wifi interface is active then it's not possible to access to the network. Activating the eth interface makes accessing the network work.

The only difference with the previous bug is that I'm using NetworkManager to manage the wifi interface.

Below are my .network and .netdev files:

```
# cat 30-bond0-eth.network
[Match]
Name=enp0s31f6

[Network]
Bond=bond1
# Only meaningful when the bond mode is "active-backup"
PrimarySlave=true
```

```
# cat 30-bond0-wifi.network 
[Match]
Name=wlp0s20f3

[Network]
Bond=bond1
```

```
# cat 30-bond0.netdev
[NetDev]
Name=bond1
Kind=bond

[Bond]
Mode=active-backup
MIIMonitorSec=1s

```
```
# cat 30-bond0.network
[Match]
Name=bond1

[Network]
DHCP=ipv4
Domains=~.
```

This is the status of the bond interface when only the wifi interface is configured.

```
# networkctl 
IDX LINK      TYPE     OPERATIONAL SETUP      
  1 lo        loopback carrier     unmanaged  
  4 enp0s31f6 ether    no-carrier  configuring
  5 wlp0s20f3 wlan     enslaved    configured 
 20 bond1     bond     degraded    configuring
```

```
 # networkctl status bond1 
● 20: bond1                                                                
                     Link File: /usr/lib/systemd/network/99-default.link   
                  Network File: /etc/systemd/network/30-bond0.network      
                          Type: bond                                       
                         State: degraded-carrier (configured) 
                        Driver: bonding                                    
                    HW Address: da:0c:10:64:a8:0a                          
                           MTU: 1500 (min: 68, max: 65535)                 
                         QDisc: noqueue                                    
  IPv6 Address Generation Mode: eui64                                      
                          Mode: active-backup                              
                        Miimon: 10s                                        
                       Updelay: 0                                          
                     Downdelay: 0                                          
          Queue Length (Tx/Rx): 16/16                                      
              Auto negotiation: no                                         
                         Speed: n/a                                        
                       Address: 2a01:e34:eea0:c810:d80c:10ff:fe64:a80a     
                                fe80::d80c:10ff:fe64:a80a                  
                       Gateway: fe80::2266:cfff:fe19:87d0                  
                           DNS: fd0f:ee:b0::1                              
                 Route Domains: .                                          
             DHCP6 Client DUID: DUID-EN/Vendor:0000ab114b0c889215d079950000

Mar 22 22:03:30 pluto systemd-networkd[28509]: bond1: Setting NDisc routes.
Mar 22 22:03:30 pluto systemd-networkd[28509]: bond1: State changed: configured -> configuring
Mar 22 22:03:30 pluto systemd-networkd[28509]: bond1: NDisc routes set.
Mar 22 22:03:30 pluto systemd-networkd[28509]: bond1: link_check_ready(): static routes are not configured.
Mar 22 22:03:30 pluto systemd-networkd[28509]: bond1: Remembering updated address: 2a01:e34:eea0:c810:d80c:>
Mar 22 22:03:30 pluto systemd-networkd[28509]: bond1: link_check_ready(): static routes are not configured.
Mar 22 22:03:30 pluto systemd-networkd[28509]: bond1: NDisc SLAAC addresses set.
Mar 22 22:03:30 pluto systemd-networkd[28509]: bond1: Removing old NDisc information obtained from fe80::22>
Mar 22 22:03:30 pluto systemd-networkd[28509]: bond1: link_check_ready(): dhcp4:no dhcp6_addresses:no dhcp_>
Mar 22 22:03:30 pluto systemd-networkd[28509]: bond1: State changed: configuring -> configured

```

@yuwata asked me to open a new bug report.

Thanks in advance.

**Comment by fmoreau** [comment]
Please also find the debug logs when systemd-networkd is restarted: [journal.txt](https://github.com/systemd/systemd/files/8328025/journal.txt)

**Comment by yuwata** [comment]
Hm,,, you use both networkd and NetworkManager to manager the wlan interface... It is not supported in general.

Also, the attached log contains mostly only after restarting networkd. Could you provide the failed case?

Also, please provide the version of systemd.

**Comment by fmoreau** [comment]
I didn't know that networkd was managing the wifi interface. I thought that networkd was only used for bonding the wifi interface. If that's not supported how should the wifi interface be setup ? because in my understanding networkd is not able to deal with wifi interfaces, no ?

The attached logs is indeed after restarting networkd. But after restarting I have no access to the network anymore so it's the failing case. Or do you want to see something else ?

I'm using systemd version 246.

**Comment by fmoreau** [comment]
Hello @yuwata , do you need any other information from me that could help ? thank you.

**Comment by yuwata** [comment]
@fmoreau sorry for late response.

> I didn't know that networkd was managing the wifi interface. I thought that networkd was only used for bonding the wifi interface. If that's not supported how should the wifi interface be setup ? because in my understanding networkd is not able to deal with wifi interfaces, no ?

networkd does not authenticate for connecting to APs, so you need to also use wpa_supplicant or iwd to do that (IIUC, NetworkManager also does not that and it collaborates with wpa_supplicant or so).
Other part of managing WLAN interfaces, such as assigning addresses, routes, etc, running DHCP client or friends, and joining a bond or bridge, are fully supported by networkd as the same as wired connection.
So, only extra setup for using WLAN interfaces with networkd is, just configure and start wpa_supplicant or friends.

One benefit to use NetworkManager for managing WLAN interfaces is, it has graphical tools or applets to manage WLAN APs, but I guess such tools just configure wpa_supplicant or so. If you feel OK to edit wpa_supplicant.conf by an editor each time you connect a new WLAN AP, then you can fully disable NetworkManager and switch to networkd.

> The attached logs is indeed after restarting networkd. But after restarting I have no access to the network anymore so it's the failing case. Or do you want to see something else ?

I have no idea. I guess there exist some inferences between networkd and NetworkManager. If so, that's not supported, sorry.
Please try to run only networkd (or only NetworkManager).

**Comment by yuwata** [comment]
> I'm using systemd version 246.

Let's close this at least tentatively, as v246 is too old here. And also tag this 'not-supported'.

@fmoreau Feel free to request to reopen this if you still have the issue even if you use only networkd from recent release of systemd.

**Comment by fmoreau** [comment]
> networkd does not authenticate for connecting to APs, so you need to also use wpa_supplicant or iwd to do that (IIUC, NetworkManager also does not that and it collaborates with wpa_supplicant or so). Other part of managing WLAN interfaces, such as assigning addresses, routes, etc, running DHCP client or friends, and joining a bond or bridge, are fully supported by networkd as the same as wired connection. So, only extra setup for using WLAN interfaces with networkd is, just configure and start wpa_supplicant or friends.
> 
> One benefit to use NetworkManager for managing WLAN interfaces is, it has graphical tools or applets to manage WLAN APs, but I guess such tools just configure wpa_supplicant or so. If you feel OK to edit wpa_supplicant.conf by an editor each time you connect a new WLAN AP, then you can fully disable NetworkManager and switch to networkd.
 
yes using NetworkManager is much more convenient to use to connect to different APs when traveling. NM is only used to connect to AP, the rest is normally done by networkd in my case:  networkd creates the bond and assign the wlan interface to the bond and run all DHCP, etc..

**Comment by fmoreau** [comment]
FYI I tried to use wpa_supplicant and same issue.

I rebooted the machine and it magically started to work with NM. So it seems that there's something fishy here.

**Comment by blatte** [comment]
I'm having the exact same issue. There's no connectivity when on wireless (which works well otherwise without bonding). Ethernet works just fine. From my various tests, I noticed that when starting with wifi, the bond interface does not get an IP address (I'm using DHCP). When I start the ethernet interface, the bond interface gets an IP and connectivity is up. Unplugging ethernet again, the bond still has its IP address but no connectivity at all through wireless.

So I wonder if the problem could be related to the wifi not getting an IP and setting routes properly for the bond interface.

**Comment by yuwata** [comment]
@blatte If you use iwd for only connecting AP, that is, managing IP address or so is done by networkd, then please open a new issue page with debugging log of networkd.
